compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2833,11 +2834,20 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       case rhs =>
         excludeDeferredGiven(rhs, sym):
           typedExpr(_, tpt1.tpe.widenExpr)
-    val vdef1 = assignType(cpy.ValDef(vdef)(name, tpt1, rhs1), sym)
-    postProcessInfo(vdef1, sym)
-    vdef1.setDefTree
+    setAbstractTrackedInfo(sym, rhs1, tpt)
+    val tpt2 = if sym.flags.is(Tracked) && tpt.isEmpty && !sym.flags.is(ParamAccessor) && !sym.flags.is(Param) then TypeTree(rhs1.tpe) else tpt1
+    val vdef2 = assignType(cpy.ValDef(vdef)(name, tpt2, rhs1), sym)
+    postProcessInfo(vdef2, sym)
+    vdef2.setDefTree
   }
 
+  private def setAbstractTrackedInfo(sym: Symbol, rhs: Tree, tpt: untpd.Tree)(using Context): Unit =
+    if !sym.flags.is(ParamAccessor) && !sym.flags.is(Param) then
+      if sym.allOverriddenSymbols.exists(_.flags.is(Tracked)) then
+        sym.setFlag(Tracked)
+      if sym.flags.is(Tracked) && tpt.isEmpty then
+        sym.info = rhs.tpe
```

Comment:
It seems dangerous to set `info` here. This means that the symbol info gets modified when the `val` is type-checked, so it's not the same before and after in definition order. 

Here is an example:

```scala
abstract class Vec:
  tracked val size: Int

@main def main =
  val v = new Vec:
    val size0: size.type = 10 // error: found (10 : Int), required (size : Int)
    val size = 10
    val size1: size.type = 10 // okay, because size is known to be (10 : Int) here
```

I am wondering if the logic wouldn't be better placed in `Namer.Completer`, maybe setting the flag in `completeInCreationContext` and computing the precise member type in `valOrDefDefSig`? Have you thought about doing so? However this is not easy to do without causing cyclic references.

Reply:
I haven't been able to do exactly what I described above due to cyclic references, but I could move the whole `setAbstractTrackedInfo` to `Namer.Completer`: https://github.com/dotty-staging/dotty/pull/53.
Still seems dirty to write over `.info`, but at least it's done directly when completing the symbol. What do you think?

Reply:
Yes, seemed dirty to me too, but it also seemed like the least intrusive way to do it. If moving the logic to Namer works, then it should be the better solution, so we can merge https://github.com/dotty-staging/dotty/pull/53.
I would probably wait on Martin's opinion though. What do you think?

Reply:
Sounds good to me to wait on Martin's opinion.
And my PR currently fails picklings test actually.